pallets/dactr/src/extensions/check_app_id.rs

```diff
@@ -27,49 +32,86 @@ use crate::{Call as DACall, Config, Pallet};
+	pub AppId,
+	sp_std::marker::PhantomData<T>,
+);
 
 impl<T> CheckAppId<T>
 where
-	T: Config + Send + Sync,
-	T::RuntimeCall: IsSubType<DACall<T>>,
+	T: DAConfig + UtilityConfig + Send + Sync,
+	<T as SystemConfig>::RuntimeCall: IsSubType<DACall<T>> + IsSubType<UtilityCall<T>>,
 {
 	/// utility constructor. Used only in client/factory code.
 	pub fn from(app_id: AppId) -> Self { Self(app_id, sp_std::marker::PhantomData) }
 
+	/// It validates that `AppId` is correct and already registered for the call and potential nested calls.
 	/// Transaction validation:
-	///  - Only `DataAvailability::submit_data(..)` extrinsic can use `AppId != 0`. Any other call
-	///  must use `AppId == 0`.
-	///  - It validates that `AppId` is already registered.
-	pub fn do_validate(&self, call: &T::RuntimeCall) -> TransactionValidity {
-		match call.is_sub_type() {
-			// Only `dactrl::submit_data` can use `AppId != 0`.
-			Some(DACall::<T>::submit_data { .. }) => {
-				let next_app_id = <Pallet<T>>::peek_next_application_id();
+	///  - `DataAvailability::submit_data(..)` extrinsic can use `AppId != 0`.
+	///  - `Utility::batch/batch_all/force_batch(..)` extrinsic can use `AppId != 0` If the wrapped calls are ALL `DataAvailability::submit_data(..)`.
+	///  - Any other call must use `AppId == 0`.
+	pub fn do_validate(&self, call: &<T as SystemConfig>::RuntimeCall) -> TransactionValidity {
+		if self.app_id() == AppId(0) {
+			return Ok(ValidTransaction::default());
+		}
+
+		let mut stack = Vec::new();
+		stack.push(call);
+
+		let mut maybe_next_app_id: Option<AppId> = None;
+		let mut iterations = 0;
+
+		while let Some(call) = stack.pop() {
+			if let Some(DACall::<T>::submit_data { .. }) = call.is_sub_type() {
+				let next_app_id =
+					maybe_next_app_id.unwrap_or_else(|| <Pallet<T>>::peek_next_application_id());
 				ensure!(
 					self.app_id() < next_app_id,
 					InvalidTransaction::Custom(InvalidTransactionCustomId::InvalidAppId as u8)
 				);
-			},
-			_ => {
-				// Any other call must use `AppId == 0`.
-				ensure!(
-					self.app_id().0 == 0,
-					InvalidTransaction::Custom(InvalidTransactionCustomId::ForbiddenAppId as u8)
-				);
-			},
-		};
+				maybe_next_app_id = Some(next_app_id);
+			} else {
+				match call.is_sub_type() {
+					Some(UtilityCall::<T>::batch { calls })
+					| Some(UtilityCall::<T>::batch_all { calls })
+					| Some(UtilityCall::<T>::force_batch { calls }) => {
+						iterations += 1;
+						ensure!(
+							iterations < MAX_ITERATIONS,
+							// TODO @Marko @Ghali Error name - Replace with MaxRecursionExceeded when PR gets merged
+							InvalidTransaction::Custom(
+								InvalidTransactionCustomId::ForbiddenAppId as u8
+							)
+						);
+						for call in calls.iter() {
+							// TODO @Marko Make it readable
+							let a = call as *const <T as UtilityConfig>::RuntimeCall;
+							let b = a as *const <T as SystemConfig>::RuntimeCall;
+							let c = unsafe { &*b };
```

Comment:
I would not use this kind of casting, because it depends on how `RuntimeCall` is defined, and that is generated by the specific version of substrate. I mean, a future change in substrate version could break this.

Reply:
Removed unsafe code. Utility::RuntimeCall implements trait IsType<<Self as frame_system::Config>::RuntimeCall> which means that is trivial to convert to and back from < frame_system::Config>::RuntimeCall